crypto/kdf/x942kdf.c

```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright (c) 2019, Oracle and/or its affiliates.  All rights reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdlib.h>
+#include <stdarg.h>
+#include <string.h>
+#include <openssl/hmac.h>
+#include <openssl/evp.h>
+#include <openssl/kdf.h>
+#include "internal/cryptlib.h"
+#include "internal/evp_int.h"
+#include "kdf_local.h"
+
+struct evp_kdf_impl_st {
+    const EVP_MD *md;
+    unsigned char *secret;
+    size_t secret_len;
+    unsigned char *other; /* DER encoded other info */
+    size_t other_len;
+    size_t ctr_offset; /* offset into the 'other' info */
+};
+
+#define X942KDF_MAX_INLEN (1<<30)
+
+static int x942_hash_kdm(const EVP_MD *kdf_md,
+                         const unsigned char *z, size_t z_len,
+                         const unsigned char *other, size_t other_len,
+                         unsigned char *ctr,
+                         unsigned char *derived_key, size_t derived_key_len)
+{
+    int ret = 0, hlen;
+    size_t counter, out_len, len = derived_key_len;
+    unsigned char mac[EVP_MAX_MD_SIZE];
+    unsigned char *out = derived_key;
+    EVP_MD_CTX *ctx = NULL, *ctx_init = NULL;
+
+    if (z_len > X942KDF_MAX_INLEN || other_len > X942KDF_MAX_INLEN
+            || derived_key_len > X942KDF_MAX_INLEN
+            || derived_key_len == 0)
+        return 0;
+
+    hlen = EVP_MD_size(kdf_md);
+    if (hlen <= 0)
+        return 0;
+    out_len = (size_t)hlen;
+
+    ctx = EVP_MD_CTX_create();
+    ctx_init = EVP_MD_CTX_create();
+    if (ctx == NULL || ctx_init == NULL)
+        goto end;
+
+    if (!EVP_DigestInit(ctx_init, kdf_md))
+        goto end;
+
+    for (counter = 1;; counter++) {
+        /* updating the ctr modifies 4 bytes in the 'other' buffer */
+        ctr[0] = (unsigned char)((counter >> 24) & 0xff);
+        ctr[1] = (unsigned char)((counter >> 16) & 0xff);
+        ctr[2] = (unsigned char)((counter >> 8) & 0xff);
+        ctr[3] = (unsigned char)(counter & 0xff);
+
+        if (!(EVP_MD_CTX_copy_ex(ctx, ctx_init)
+                && EVP_DigestUpdate(ctx, z, z_len)
+                && EVP_DigestUpdate(ctx, other, other_len)))
```

Comment:
Personally I find the style of `!(x && y && z)` as slightly jarring and different to what we typically do which is `!x || !y || !z`

Reply:
I find it easier to understand. It basically says if any of these fail.. I will conform if you really want.

Reply:
I don't feel strongly enough about it to block the progress of this PR - but at some point we do need to have some discussions on coding style. There are a whole load of things that we currently do (and enforce in PR reviews) which aren't captured by the coding style doc.

Reply:
_(@slontis your viewpoint is interesting, because for me it's just the other way around: if I translate "**if any** of these fail" into boolean algebra, I get "**if** x fails **or** y fails **or** z fails".)_

No matter which one of `!(x && y && z)` or  `!x || !y || !z` is used, there will always be somebody who finds the other variant more intuitive. Sometimes it even depends on the situation. So I'd let statistics decide and choose @mattcaswell' variant for consistency and better readability.

Reply:
Maybe some more rules in the coding style would be good.
Sometimes I still dont know what to do with line wrapping, and indentation after a line wrap does vary.

Reply:
Changed